docs/configuration.md
# Configuration Options

All CloudPunch configuration is given through the command-line

## Command-line Options

The following options are given on the command-line when using cloudpunch

- `workload` - The type of workload to run. Can be run, cleanup, post, or worker. These are covered below

- `-h, --help` - Show the help message and exit

- `-v, --version` - Show the version number and exit

- `-l, --loglevel` - Specify a log level when the default of INFO is not wanted. The following log levels are valid: DEBUG, INFO, WARNING, ERROR, and CRITICAL

- `-L, --logfile` - Specify a log file to send all logging to rather than `stdout`. Note that this will overwrite any file given and **not** append

## Run Command-line Options

The following options are given on the command-line when using cloudpunch run

- `-c, --config` - Provide a JSON formatted configuration file that will override the default configuration. Any keys that are missing will use the default values. See Configuration Files below for more information

- `-e, --env` - Provide a JSON formatted environment file that will override the default environment configuration. Any keys that are missing will use the default values. See Environment Files below for more information

- `-e2, --env2` - Environment file used for the second OpenStack instance if `--split` is used

- `-r, --openrc` - Provide an OpenRC file containing authentication information for OpenStack. Note that information in the environment will override anything in the OpenRC file

- `-r2, --openrc2`- OpenRC file for second OpenStack instance if `--split` is used

- `-m, --hostmap` - Provide a hostmap file to determine where each instance created will be hosted on. See Hostmap Files below for more information

- `-b, --flavor` - Provide a flavor file to determine the percentage of instances with specific flavors. This will override any flavor configuration. See Flavor Files below for more information

- `-o, --output` - Specify an output file to save results to. Results will be printed to `stdout` if an output file is not specified

- `-f, --format` - Format the results into the specified format. Can be yaml or json with yaml being the default

- `-p, --password` - Provide a password or token to authenticate to OpenStack. This should only be used when in a non-interactive environment and the environment does not contain the password or token

- `-p2, --password2` - Password for second OpenStack instance if `--split` is used

- `-i, --reuse` - Enable reuse mode by providing an existing CloudPunch run ID. This will reuse any resources that belong to this ID and adjust the environment to match the new configuration

- `-l, --listen` - Local address that the control server will bind to. The default is 0.0.0.0

- `-t, --port` - Local port that the control server will bind to. The default is 9985. This will populate down to the worker servers

- `-w, --connect` - Provide a local interface or IP address as a hint to what the worker servers on OpenStack will connect back to when communicating with the local control server. By default CloudPunch will attempt to find an IP address based on the default gateway of the local machine. Note that this does not have to be a known IP address or DNS name from the local machine (for example: a floating IP address on OpenStack)

- `--no-env` - Disables loading authentication information from the environment. Use this to force the OpenRC file over the environment

- `--manual` - Enable manual test start mode. After the environment is staged and ready but before the test begins, the user must press Enter to continue. Note that this requires interactive

- `--insecure` - Turn off SSL verification to the OpenStack API's

## Cleanup Command-line Options

The following options are given on the command-line when using cloudpunch cleanup

- `cleanup_file` - Provide the cleanup file containing all resource IDs. This cleanup file is created when CloudPunch terminates. This can be 'search' to search for resources that have been left over

- `-r, --openrc` - Provide an OpenRC file containing authentication information for OpenStack. Note that information in the environment will override anything in the OpenRC file

- `-p, --password` - Provide a password to authenticate to OpenStack. This should only be used when in a non-interactive environment and the environment does not contain the password

- `-n, --dry-run` - Do not delete after searching for resources

- `-a, --names` - Show names and UUIDs of found resources

- `--no-env` - Disables loading authentication information from the environment. Use this to force the OpenRC file over the environment

- `--insecure` - Turn off SSL verification to the OpenStack API's

## Post Command-line Options

The following options are given on the command-line when using cloudpunch post

- `results_file` - Provide the results file from a CloudPunch run

- `-f, --format` - The format to convert the processed results to. Can be json, yaml (default), table, or csv

- `o, --output` - Specify an output file to save processed results to

- `-s, --stat` - Stat from test to graph (graph format only)

- `-t, --test` - Name of the CloudPunch test to graph (graph format only). By default all tests will be graphed

- `-j, --job` - Name of the FIO job to graph (FIO test and graph format only)

- `--summary` - Convert over time results to summary results

- `--raw` - Do not convert numbers to human readable format. By default 3000 would be converted to 3 K

- `--open` - Open the generated HTML graph file after creation (graph format only)

## Worker Command-line Options

The following options are given on the command-line when using cloudpunch worker

- `control_ip` - IP address of the control server

- `-p, --port` - Port of the control server. The default is 9985

## Configuration File

The configuration is a JSON or YAML formatted file containing information needed to run CloudPunch and is provided using the `-c` command-line option. Default values will be used if keys are missing from the file. The configuration file is exposed to the tests running and often will contain extra information needed by the specific test. See the documentation for the specific test for information on what extra keys are required.

##### Default Configuration

```yaml
cleanup_resources: true
server_client_mode: true
servers_give_results: true
overtime_results: false
instance_threads: 5
retry_count: 50
network_mode: full
number_routers: 1
networks_per_router: 1
instances_per_network: 1
test:
  - ping
test_mode: list
test_start_delay: 0
metrics:
  enable: false
  topic:
  brokers: []
  format: influxdb
  tags: {}
```

##### Configuration Key Reference

- `cleanup_resources` - If resources should be deleted at the end of the process. This is on all cases including if an error occurs. If `false`, a cleanup file will be created in the current directory. This cleanup file is used with `cleanup.py` to delete the environment at a later time

- `server_client_mode` - If the test requires that there is a server and a client. This will double the number of routers, networks, and instances created. The `role` key is used inside tests to know if an instance is server or client. The default role is server

- `servers_give_results` - If the role of server gives results to the test. Note that this is ignored if `server_client_mode` is disabled. This option also includes if clients give results as it only causes the number of instances expected to give results to be half of the total number of instances

- `overtime_results` - If the results given from tests should be overtime or a summary. For example, should iperf give results every interval or should it give an average bps for the whole test. Note that tests will have to implement this configuration

- `instance_threads` - How many instances should be created at the same time. Adjust this number based on the power of the OpenStack environment

- `retry_count` - Number of retries done before a process is considered a failure. This is used during creation and connecting to instances

- `network_mode` - The type of network setup for the instances. The following options are allowed:

  - "full" - Create routers, networks, and instances with floating IP addresses

  - "single-router" - Create networks and instances without floating IP addresses. All networks will be connected to the master router and use internal IP addresses to communicate. Note that this requires SNAT

  - "single-network" - Create instances with floating IP addresses. All instances will be connected to the master network and use internal IP addresses to communicate. Note that this requires SNAT

- `number_routers` - The number of routers to be created. This is only used if `network_mode` is "full"

- `networks_per_router` - The number of networks to be created for each router. This is ignored if `network_mode` is "single-network"

- `instances_per_network` - The number of instances to be created for each network

- `test` - A list of test names to run. These test names are derived from the filenames inside `worker/`. For example `iperf.py` would be "iperf"

- `test_mode` - The type of test processing that should occur. The following options are allowed:

  - "list" - Run the tests in order where only one runs at a time

  - "concurrent" - Run the tests at the same time

- `test_start_delay` - Number of seconds to wait before a test starts. If `test_mode` is "list" the delay will be applied before the start of each test. For example: wait, test, wait, test. If `test_mode` is "concurrent" the delay will be applied only before the initial start. For example: wait, all tests

- `metrics` - Used to send test metrics to Kafka while tests are running. `metrics` has the following sub keys:

  - `enable` - If to enable sending metrics

  - `topic` - Name of the Kafka topic to send messages to

  - `brokers` - A list of Kafka brokers to connect to

  - `format` - The format of the messages sent to Kafka. Currently only influxdb is supported

  - `tags` - A dictionary of tags to add to all metrics sent to Kafka

## Environment Files

The environment is a JSON or YAML formatted file containing information needed to run CloudPunch in specific environments and is provided using the `-e` command-line option. Default values will be used if keys are missing from the file

##### Default Environment

```yaml
image_name: CentOS7
public_key_file: ~/.ssh/id_rsa.pub
api_versions:
  cinder: 2
  glance: 2
  nova: 2
  neutron: 2
  lbaas: 2
server:
  flavor: m1.small
  availability_zone:
  volume:
    enable: false
    size: 10
    type:
    availability_zone:
  boot_from_vol:
    enable: false
    size: 10
  loadbalancer:
    enable: false
    method: ROUND_ROBIN
    frontend:
      protocol: HTTP
      port: 80
    backend:
      protocol: HTTP
      port: 80
    healthmonitor:
      type: PING
      delay: 5
      timeout: 5
      retires: 3
      url_path: /
      http_method: GET
      expected_codes: "200"
  userdata:
client:
  flavor: m1.small
  availability_zone:
  volume:
    enable: false
    size: 10
    type:
    availability_zone:
  boot_from_vol:
    enable: false
    size: 10
  loadbalancer:
    enable: false
    method: ROUND_ROBIN
    frontend:
      protocol: HTTP
      port: 80
    backend:
      protocol: HTTP
      port: 80
    healthmonitor:
      type: PING
      delay: 5
      timeout: 5
      retires: 3
      url_path: /
      http_method: GET
      expected_codes: "200"
  userdata:
secgroup_rules:
  - - icmp
    - -1
    - -1
  - - tcp
    - 22
    - 22
  - - tcp
    - 5201
    - 5201
dns_nameservers:
  - 8.8.8.8
  - 8.8.4.4
shared_userdata:
  - mkdir -p /opt/cloudpunch
  - git clone https://github.com/target/cloudpunch.git /opt/cloudpunch
  - cd /opt/cloudpunch
  - python setup.py install
external_network:
```

##### Environment Key Reference

- `image_name` - The OpenStack image name used when creating all instances. This can be a name or ID

- `public_key_file` - Path to the public key to use when creating a keypair. This is used to allow SSH access for troubleshooting purposes

- `api_versions` - The OpenStack API versions to use when creating and deleting resources. Note that the Keystone version is determined by authentication parameters. `api_versions` has the following sub keys:

  - `cinder` - The version of cinder (storage) to use

  - `nova` - The version of nova (compute) to use

  - `neutron` - The version of neutron (network) to use

  - `lbaas` - The version of lbaas (loadbalancer as a service) to use

- `server` - Properties that apply to the server role. The server role is a worker that is designated as a server during creation and when tests are running. `server` has the following sub keys:

  - `flavor` - The name of the flavor to use when creating the instance. This must be the name and not an ID

  - `availability_zone` - The availability zone to attach the server role to

  - `volume` - Properties that apply to an attached cinder volume. Note that CloudPunch does not format the volume and it is recommended to use userdata to format and mount the volume. `volume` has the following sub keys:

      - `enable` - If a volume should be created and attached

      - `size` - The size of the volume in Gigabytes

      - `type` - The type of the volume. This is not required

      - `availability_zone` - The availability zone of the volume. This is not required

  - `boot_from_vol` - Properties that apply to booting from a volume rather than a local storage. `boot_from_vol` has the following sub keys:

      - `enable` - If the instance should boot from a volume rather than local storage

      - `size` - The size of the volume in Gigabytes

  - `loadbalancer` - Properties that apply to incorporating a loadbalancer in front of servers. `loadbalancer` has the following sub keys:

      - `enable` - If to enable the loadbalancer creation

      - `method` - The loadbalancer algorithm. Valid options are ROUND_ROBIN, LEAST_CONNECTIONS, or SOURCE_IP

      - `frontend` - Properties that apply to the frontend of the loadbalancer. `frontend` has the following sub keys:

          - `protocol` - Type of traffic protocol. Valid options are HTTP, HTTPS, or TCP

          - `port` - The protocol port

      - `backend` - Properties that apply to the backend of the loadbalancer. `backend` has the following sub keys:

          - `protocol` - Type of traffic protocol. Valid options are HTTP, HTTPS, or TCP

          - `port` - The protocol port

      - `healthmonitor` - Properties that apply to the loadbalancer's health monitor

          - `type` - The health monitor type. Valid options are PING, TCP, HTTP or HTTPS

          - `delay` - The delay between checks in seconds

          - `timeout` - The timeout period in seconds

          - `retires` - The maximum number of retries before a member is removed. Must be between 1 and 10

          - `url_path` - The url path to test. Only valid with HTTP or HTTPS types

          - `http_method` - The HTTP method to use to test. Only valid with HTTP or HTTPS types

          - `expected_codes` - A single (200), range (200-202), or comma-seperated (200,201,202) list of codes to consider valid

  - `userdata` - A list of commands processed by shell to run during the cloud-init process. This is used to setup the server role for specific environments and tests

- `client` - Properties that apply to the client role. The client role is a worker that is designated as a client during creation and when tests are running. `client` has the following sub keys:

  - `flavor` - The name of the flavor to use when creating the instance. This must be the name and not an ID

  - `availability_zone` - The availability zone to attach the client role to

  - `volume` - Properties that apply to an attached cinder volume. Note that CloudPunch does not format the volume and it is recommended to use userdata to format and mount the volume. `volume` has the following sub keys:

      - `enable` - If a volume should be created and attached

      - `size` - The size of the volume in Gigabytes

      - `type` - The type of the volume. This is not required

      - `availability_zone` - The availability zone of the volume. This is not required

  - `boot_from_vol` - Properties that apply to booting from a volume rather than a local storage. `boot_from_vol` has the following sub keys:

      - `enable` - If the instance should boot from a volume rather than local storage

      - `size` - The size of the volume in Gigabytes

  - `loadbalancer` - Properties that apply to incorporating a loadbalancer in front of clients. `loadbalancer` has the following sub keys:

      - `enable` - If to enable the loadbalancer creation

      - `method` - The loadbalancer algorithm. Valid options are ROUND_ROBIN, LEAST_CONNECTIONS, or SOURCE_IP

      - `frontend` - Properties that apply to the frontend of the loadbalancer. `frontend` has the following sub keys:

          - `protocol` - Type of traffic protocol. Valid options are HTTP, HTTPS, or TCP

          - `port` - The protocol port

      - `backend` - Properties that apply to the backend of the loadbalancer. `backend` has the following sub keys:

          - `protocol` - Type of traffic protocol. Valid options are HTTP, HTTPS, or TCP

          - `port` - The protocol port

      - `healthmonitor` - Properties that apply to the loadbalancer's health monitor

          - `type` - The health monitor type. Valid options are PING, TCP, HTTP or HTTPS

          - `delay` - The delay between checks in seconds

          - `timeout` - The timeout period in seconds

          - `retires` - The maximum number of retries before a member is removed. Must be between 1 and 10

          - `url_path` - The url path to test. Only valid with HTTP or HTTPS types

          - `http_method` - The HTTP method to use to test. Only valid with HTTP or HTTPS types

          - `expected_codes` - A single (200), range (200-202), or comma-seperated (200,201,202) list of codes to consider valid

  - `userdata` - A list of commands processed by shell to run during the cloud-init process. This is used to setup the client role for specific environments and tests

- `secgroup_rules` - A list of security group rules that will be attached to all instances created. Each rule is a list containing protocol, from port, and to port. For example: `["icmp", -1, -1]`

- `dns_nameservers` - A list of DNS nameservers that will be added to all subnets created

- `shared_userdata` - A list of commands processed by shell to run during the cloud-init process. This is added **before** each others userdata

- `external_network` - The name of the external network to connect routers and assign floating IP addresses from. If no name is given, the first result from the list of external networks will be used


## Hostmap Files

The hostmap is a JSON or YAML formatted file containing a list of availability zones or hosts that will determine where each instances will land. Note that when using a hostmap, the `availability_zone` keys for servers and clients will be ignored. The master key will not be ignored. It consists of two main keys: `tags` and `map`. `tags` is used to translate something in the map to it's real name. For example `compute-zone1` could be tagged as `zone1`. The `map` key is a list of availability zones or tags. Each entry has two comma separated availability zones or tags. The first is where the server will end up and the second is where the client will end up. This pair of server and client will be associated with each other later on for tests. This list is processed in order of creation and if the list is exhausted, it will begin at the beginning again.

##### Hostmap Example

```yaml
tags:
  zone1: compute-zone1
  zone2: compute-zone2
  zone3: compute-zone3
  zone4: compute-zone4
  zone5: compute-zone5
  zone6: compute-zone6
map:
  - zone1,zone2
  - zone2,zone3
  - zone3,zone4
  - zone4,zone5
  - zone5,zone6
  - zone6,zone1
```

## Flavor Files

The flavor file is a JSON or YAML formatted file containing a list of flavors assigned to a percentage. This will determine how many instances will be assigned a specific flavor. Note that when using a flavor file, the `flavor` keys for servers and clients will be ignore. The master key will not be ignored. If `server_client_mode` is enabled, both servers and clients will follow the flavor file the same way. Note that the percentages have to add up to be between 99% and 100%. This is to account for scenarios that are close to 100% but not exact such as 33.33% or 1/3.

##### Flavor File Example

```yaml
flavors:
  m1.tiny: 0
  m1.demo: 0
  m1.small: 50
  m1.medium: 50
  m1.large: 0

CHANGELOG.md
# Change Log
A list of notable changes will be documented here

## 2.0.0 - 2017-xx-xx
### Added
- A new parameter `-l, --listen` has been added to `cloudpunch run`. This configures the binding address of the control server
- A new parameter `-t, --port` has been added to `cloudpunch run`. This configures the port that the control server listens on
- A new parameter `-w, --connect` has been added to `cloudpunch run`. This configures how the workers connect to the local control server. The supplied value is either an IP address or interface name
- All official tests now include sending metrics to Kafka. Currently the only supported format is influxdb. See documentation for more information
- Doing cleanup search will now search for load balancers v1 and v2

### Changed
- The master server has been removed and now is run as a control server on the local machine
- The control server will only respond to workers that have the same cloudpunch run id
- Slaves are now called workers. The `cloudpunch slave` command is now `cloudpunch worker`
- Fixed cloudpunch cleanup search looping through floating ip addresses multiple times
- All official tests now have normalized settings and results. See documentation for specifics
- All graphs now use datetime instead of the number of seconds for the X axis
- The ping test now has time as an integer rather than a float to match other tests
- The stress test now has time as the current time rather than duration
- The jmeter test now has time as the current time rather than the test duration to match other tests
- The run command now uses `-f FORMAT` instead of `--yaml`. Yaml is now the default option
- The run command uses `-b FLAVOR_FILE` instead of `-f` for providing a flavor breakdown

### Removed
- Recovery is no longer an option
- The command `cloudpunch master` has been removed. The master server is now run as a control server during the `cloudpunch run` command
- Admin mode has been removed, CloudPunch will now only use the current user and project
- The run command no longer has `--split`. Supplying a second OpenRC file will enable split mode
- The run command no longer has `--reuse`. The new control server design has removed this ability

## 1.5.0 - 2017-08-16
### Added
- New configuration options in the jmeter test. You can now specify `port` and `path` when connecting to a server
- The jmeter test now supports `overtime_results`
- stress-ng test now supports summary results
- New `graph` format option in cloudpunch post. This option uses plotly to graph results into an HTML file
- New `--summary` option in cloudpunch post. This option will convert over time results to summary results

### Changed
- Fixed incorrect overtime results in the FIO test
- ping test no longer includes target and duration

## 1.4.0 - 2017-06-14
### Added
- Cleanup now has the flag `-n, --dry-run` when performing a search. This will prevent resources from being deleted
- Cleanup now has the flag `-a, --names` when performing a search. It will print out the UUIDs and names of found resources

### Changed
- The SSH public key file specified in the environment file is now checked if it exists
- Security groups are now made through Neutron instead of Nova (deprecated in novaclient 8)
- Images are now searched inside Glance instead of Nova (deprecated in novaclient 9)
- Cleanup will now only stay within the sourced project when an admin user
- Cleanup will now not list out repeated neutron resources

## 1.3.1 - 2017-05-01
### Added
- availability_zone is now an option under server and client volume

### Changed
- Babel is now pinned to version 2.3.4 to avoid dependency issues

## 1.3.0 - 2017-04-20
### Added
- Load balancer support in front of servers and/or clients. See Configuration documentation for more information
- Custom test files can now be added into a `tests/` directory inside the current working directory. Any test files inside the configuration will be pushed over to the slaves to run
- cloudpunch post now supports formatting results to json, yaml, table, and csv
- cloudpunch post now calculates mean, median, mode, range, and total instead of just mean and total
- cloudpunch cleanup can now search for resources left over by previous cloudpunch runs. Use `cloudpunch cleanup search`

### Changed
- cloudpunch-cleanup and cloudpunch-post have now been integrated into the main cloudpunch command. To run a test use cloudpunch run, to cleanup use cloudpunch cleanup, and to post process use cloudpunch post
- cloudpunch-master and cloudpunch-slave have also been integrated into the main cloudpunch command. Use cloudpunch master to start the master and cloudpunch slave to start a slave
- All official tests now have a default configuration

## 1.2.0 - 2016-12-21
### Added
- Ability to have instances boot from volume (see environment file)
- FIO test now supports FIO test files

### Changed
- iPerf test no longer has an enable field to avoid confusion

## 1.1.1 - 2016-11-18
### Changed
- Now wait a max of 5 seconds before detaching a router interface is marked as a failure to prevent delays in neutron causing errors
- Only use the master's floating IP address when using network-mode 'full'. This allows slaves to connect to the master without the need for SNAT

## 1.1.0 - 2016-10-17
### Added
- Test results now return errors to the master if the run fails
- Left over resources after a cleanup will now be added to the cleanup file. This includes items that failed to delete
- Post cleanup now supports the insecure flag
- Post cleanup now supports multiple OpenStack API versions. The cleanup file will now include the versions for cinder, nova, and neutron
- A simple post processor for test results has been added. Run either `cloudpunch-post` or `cloudpunch/post.py`. Currently it only displays totals and averages of all instance results

### Changed

cloudpunch/configuration.py
import os
import yaml
import logging

import cloudpunch.utils.config as cpc

from cloudpunch.utils import exceptions

# List of offical files inside cp_worker (not test files)
OFFICIAL_FILES = ['__init__', 'cp_worker', 'sysinfo', 'flaskapp']


class Configuration(object):

    def __init__(self, run_id, config_file=None, output_file=None,
                 hostmap_file=None, flavor_file=None, split_mode=False):

        # Hard coded default configuration
        # This should be enough to run a test without giving a config file
        default_config = {
            'cleanup_resources': True,
            'server_client_mode': True,
            'servers_give_results': True,
            'overtime_results': False,
            'instance_threads': 5,
            'retry_count': 50,
            'network_mode': 'full',
            'number_routers': 1,
            'networks_per_router': 1,
            'instances_per_network': 1,
            'test': ['ping'],
            'test_mode': 'list',
            'test_start_delay': 0,
            'metrics': {
                'enable': False,
                'topic': '',
                'brokers': [],
                'format': 'influxdb',
                'tags': {}
            }
        }
        read_config = {}

        # Load configuration from file if specified
        if config_file:
            if not os.path.isfile(config_file):
                raise exceptions.CPError('Configuration file %s not found' % config_file)
            logging.debug('Loading configuration from file %s', config_file)
            read_config = self.loadfile(config_file, 'Configuration')
        else:
            logging.debug('Using default CloudPunch configuration')

        # Merge default and config file
        self.final_config = cpc.merge_configs(default_config, read_config)

        # Add in run id
        self.final_config['run_id'] = run_id

        # Check official tests (tests in worker/)
        worker_dir = os.path.dirname(os.path.realpath(__file__)) + '/worker'
        files = os.listdir(worker_dir)
        official_tests = []
        for filename in files:
            name, extension = os.path.splitext(filename)
            if extension == '.py' and name not in OFFICIAL_FILES:
                official_tests.append(name)
        # Check unofficial tests (tests in config)
        unofficial_tests = []
        for test in self.final_config['test']:
            if test not in official_tests:
                unofficial_tests.append(test)
        # We have unofficial tests to find in the tests directory (current-directory/tests)
        if unofficial_tests:
            self.final_config['test_files'] = {}
            test_dir = 'tests'
            for unofficial_test in unofficial_tests:
                test_file_path = '%s/%s.py' % (test_dir, unofficial_test)
                if not os.path.isfile(test_file_path):
                    raise exceptions.CPError('Unable to find unofficial test %s in tests directory' % unofficial_test)
                with open(test_file_path) as f:
                    test_file_data = f.read()
                self.final_config['test_files'][unofficial_test] = test_file_data

        # Check if server_client_mode is enabled when split mode is enabled
        if split_mode and not self.final_config['server_client_mode']:
            raise exceptions.CPError('server_client_mode is required be to enabled when split mode is enabled')
        # Split mode will only work in full network mode
        if split_mode and self.final_config['network_mode'] != 'full':
            raise exceptions.CPError('network_mode must be full when split mode is enabled')

        # Add output_file to config if specified
        if output_file:
            logging.debug('Saving results to file %s', output_file)
            self.final_config['output_file'] = output_file

        # Add hostmap to config if specified
        if hostmap_file:
            if not os.path.isfile(hostmap_file):
                raise exceptions.CPError('Hostmap file %s not found' % hostmap_file)
            logging.debug('Using hostmap file %s', hostmap_file)
            self.final_config['hostmap'] = self.loadfile(hostmap_file, 'Hostmap')

        # Add flavor file to config if specified
        if flavor_file:
            if not os.path.isfile(flavor_file):
                raise exceptions.CPError('Flavor file %s not found' % flavor_file)
            flavor_data = self.loadfile(flavor_file, 'Flavor')['flavors']
            # Remove any zero percent flavors
            for flavor in flavor_data[:]:
                if float(flavor_data[flavor]) == 0.0:
                    del flavor_data[flavor]
            # Check if total percent is 99 to 100 (99 for 1/3)
            total = 0.0
            for flavor in flavor_data:
                total += float(flavor_data[flavor])
            if total > 100.0 or total < 99.0:
                raise exceptions.CPError('Flavor file does not add up to 99-100%%. Currently %s%%' % total)
            logging.debug('Using flavor file %s', flavor_file)
            self.final_config['flavor_file'] = flavor_data

        # Add fio test_file to config if specified
        if ('fio' in self.final_config['test'] and
                'fio' in self.final_config and 'test_file' in self.final_config['fio'] and
                self.final_config['fio']['test_file']):
            test_file = self.final_config['fio']['test_file']
            if not os.path.isfile(test_file):
                raise exceptions.CPError('FIO test file %s not found' % test_file)
            with open(test_file, 'r') as f:
                test_file_data = f.read()
            self.final_config['fio']['test_file_data'] = test_file_data

        # Check numbers in configuration
        if self.final_config['instance_threads'] < 1:
            raise exceptions.CPError('Invalid number of instance_threads. Must be greater than 0')
        if self.final_config['test_start_delay'] < 0:
            raise exceptions.CPError('Invalid test_start_delay. Must be 0 or greater')
        if self.final_config['retry_count'] < 1:
            raise exceptions.CPError('Invalid retry_count. Must be greater than 0')
        if self.final_config['number_routers'] < 1:
            raise exceptions.CPError('Invalid number_routers. Must be greater than 0')
        if self.final_config['networks_per_router'] < 1:
            raise exceptions.CPError('Invalid networks_per_router. Must be greater than 0')
        if self.final_config['instances_per_network'] < 1:
            raise exceptions.CPError('Invalid instances_per_network. Must be greater than 0')

        # Check test mode
        if self.final_config['test_mode'] not in ['list', 'concurrent']:
            raise exceptions.CPError('Invalid test_mode. Must be list or concurrent')

        # Check metrics
        if self.final_config['metrics']['enable']:
            if not self.final_config['metrics']['topic']:
                raise exceptions.CPError('Missing metrics Kafka topic')
            if not self.final_config['metrics']['brokers']:
                raise exceptions.CPError('Missing metrics Kafka brokers')
            if self.final_config['metrics']['format'] != 'influxdb':
                raise exceptions.CPError('The only supported metrics format is influxdb')

        # Network and environment number checks
        if self.final_config['network_mode'] not in ['full', 'single-router', 'single-network']:
            raise exceptions.CPError('Invalid network_mode. Must be full, single-router, or single-network')
        if self.final_config['server_client_mode'] and self.final_config['number_routers'] > 126:
            raise exceptions.CPError('Number of routers cannot be greater than 126 if server_client_mode is enabled')
        if self.final_config['number_routers'] > 254:
            raise exceptions.CPError('Number of routers cannot be greater than 254')
        if self.final_config['networks_per_router'] > 254:
            raise exceptions.CPError('Number of networks per router cannot be greater than 254')
        if self.final_config['network_mode'] == 'full':
            if self.final_config['instances_per_network'] > 250:
                raise exceptions.CPError('Number of instances per network cannot be greater than 250')
        if self.final_config['network_mode'] in ['single-router', 'single-network']:
            if self.final_config['server_client_mode'] and self.final_config['networks_per_router'] > 126:
                raise exceptions.CPError('Number of networks per router cannot be greater than 126'
                                         ' if server_client_mode is enabled')
            if self.final_config['instances_per_network'] > 62500:
                raise exceptions.CPError('Number of instances per network cannot be greater than 62500'
                                         ' if network mode is single-router or single-network')

    def loadfile(self, data_file, label='Configuration'):
        with open(data_file) as f:
            contents = f.read()
        try:
            data = yaml.load(contents)
        except yaml.YAMLError as e:
            raise exceptions.CPError('%s file failed to load: %s' (label, e))
        return data

    def get_config(self):
        return self.final_config
